bin/genForest.ts

```diff
@@ -114,142 +151,361 @@ function enumerateExpressions(leaves: number): SKIExpression[] {
+      const inFlight = slots.filter(Boolean).length;
+      const slotSummary = slots
+        .map(
+          (s, i) => (s ? `${i}:${s.exprIndex + 1}@${s.stepsTaken}` : `${i}:-`),
+        )
+        .join(" ");
+      console.error(
+        `[Status] Processed: ${processed}/${total} (${
+          ((processed / total) * 100).toFixed(2)
+        }%) | In flight: ${inFlight}/${CONCURRENCY} | ` +
+          `Total time: ${elapsed}s | Slots: ${slotSummary} | ` +
+          `Ring: ${JSON.stringify(getRingStats() ?? {})} | ` +
+          `resize_seq: ${getDebugLockState() ?? "n/a"}`,
+      );
+    }, 2000);
+
+    try {
+      type Done = {
+        slot: number;
+        exprIndex: number;
+        result: EvalResult;
+      };
+      type InFlightEntry = {
+        slot: number;
+        exprIndex: number;
+        promise: Promise<Done>;
+      };
+
+      let inFlights: InFlightEntry[] = [];
+
+      const startSlot = (slot: number) => {
+        if (nextIndex >= total) {
+          return;
+        }
+        const exprIndex = nextIndex++;
+        slots[slot] = { exprIndex, stepsTaken: 0, startedAtMs: Date.now() };
+
+        const base = evaluateBatchParallel(
+          evaluator,
+          allArenaIds[exprIndex],
+          maxSteps,
+          (stepsTaken) => {
+            const s = slots[slot];
+            if (s) {
+              s.stepsTaken = stepsTaken;
+            }
+          },
+        );
+
+        const promise = base.then((result) => ({ slot, exprIndex, result }));
+        inFlights.push({ slot, exprIndex, promise });
+      };
+
+      for (let slot = 0; slot < CONCURRENCY; slot++) {
+        startSlot(slot);
       }
 
-      pathSteps.push({ from: currentId, to: nextId });
-      currentId = nextId;
-      history.push(currentId);
+      // Buffer for results to ensure deterministic output order
+      const resultBuffer = new Map<number, EvalResult>();
+      let nextExpectedIndex = 0;
+
+      while (inFlights.length > 0 || resultBuffer.size > 0) {
+        // Wait for next completion if we have in-flight work
+        if (inFlights.length > 0) {
+          const done = await Promise.race(
+            inFlights.map((e) =>
+              e.promise
+            ),
+          );
+          inFlights = inFlights.filter((e) => e.slot !== done.slot);
+
+          slots[done.slot] = null;
+          processed++;
+          sources.add(done.result.source);
+          sinks.add(done.result.sink);
+
+          // Buffer the result to maintain deterministic output order
+          resultBuffer.set(done.exprIndex, done.result);
+
+          // backfill this slot if more work remains
+          startSlot(done.slot);
+        }
+
+        // Yield results in deterministic order (by exprIndex)
+        while (resultBuffer.has(nextExpectedIndex)) {
+          const result = resultBuffer.get(nextExpectedIndex)!;
+          resultBuffer.delete(nextExpectedIndex);
+          yield JSON.stringify(result);
```

Comment:
benchmark this loop, it can turn what should be 100s of ns into 100s of micro seconds or 1-10 milliseconds